File: sidecar/src/downloader.ts
import fs from "node:fs";
import path from "node:path";
import https from "node:https";
import http from "node:http";

const UPLOAD_DIR_CANDIDATES = [
  process.env.HERMES_UPLOADS_DIR,
  "/workspace/uploads",
  path.join(process.cwd(), "cache", "uploads"),
].filter(Boolean) as string[];

function getTargetUploadDir(): string {
  for (const dir of UPLOAD_DIR_CANDIDATES) {
    try {
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
      }
      return dir;
    } catch {
      continue;
    }
  }
  const fallback = path.join(process.cwd(), "uploads");
  if (!fs.existsSync(fallback)) {
    fs.mkdirSync(fallback, { recursive: true });
  }
  return fallback;
}

export interface CachedAttachment {
  name: string;
  url: string;
  size: number;
  localPath: string;
  isText: boolean;
  textContent?: string;
  contentType?: string;
}

const TEXT_EXTENSIONS = new Set([
  ".txt", ".md", ".json", ".yaml", ".yml", ".py", ".js", ".ts", ".tsx", ".jsx",
  ".sh", ".bash", ".css", ".html", ".sql", ".env", ".toml", ".ini", ".conf", ".log", ".diff", ".patch"
]);

export async function downloadAndCacheAttachment(
  url: string,
  filename: string,
  size: number,
  contentType?: string
): Promise<CachedAttachment> {
  const uploadDir = getTargetUploadDir();
  const safeFilename = `${Date.now()}_${filename.replace(/[^a-zA-Z0-9._-]/g, "_")}`;
  const targetPath = path.join(uploadDir, safeFilename);

  await new Promise<void>((resolve, reject) => {
    const file = fs.createWriteStream(targetPath);
    const client = url.startsWith("https") ? https : http;

    client.get(url, (response) => {
      if (response.statusCode !== 200) {
        file.close();
        fs.unlink(targetPath, () => {});
        reject(new Error(`Échec du téléchargement (status: ${response.statusCode})`));
        return;
      }

      response.pipe(file);
      file.on("finish", () => {
        file.close();
        resolve();
      });
    }).on("error", (err) => {
      file.close();
      fs.unlink(targetPath, () => {});
      reject(err);
    });
  });

  const ext = path.extname(filename).toLowerCase();
  const isText = TEXT_EXTENSIONS.has(ext) || (contentType?.startsWith("text/") ?? false);
  let textContent: string | undefined = undefined;

  // Injection automatique si fichier texte < 50 Ko
  if (isText && size < 50 * 1024 && fs.existsSync(targetPath)) {
    try {
      textContent = fs.readFileSync(targetPath, "utf-8");
    } catch (err) {
      console.warn(`[downloader] Impossible de lire le texte de ${filename}:`, err);
    }
  }

  return {
    name: filename,
    url,
    size,
    localPath: targetPath,
    isText,
    textContent,
    contentType,
  };
}

File: downloader.py
"""
Module de téléchargement et de mise en cache locale des pièces jointes Discord.
Télécharge les fichiers vers le répertoire configuré (HERMES_UPLOADS_DIR),
détecte les fichiers texte/code légers (< 50 Ko) et extrait leur contenu
pour auto-injection dans le contexte de l'agent.
"""

from __future__ import annotations

import logging
import os
import re
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

import aiohttp

logger = logging.getLogger(__name__)

TEXT_EXTENSIONS = {
    ".txt", ".md", ".json", ".yaml", ".yml", ".py", ".js", ".ts", ".tsx", ".jsx",
    ".sh", ".bash", ".css", ".html", ".sql", ".env", ".toml", ".ini", ".conf",
    ".log", ".diff", ".patch", ".rs", ".go", ".c", ".cpp", ".h", ".hpp",
}

MAX_INJECT_SIZE_BYTES = 50 * 1024  # 50 Ko


@dataclass
class CachedAttachment:
    name: str
    url: str
    size: int
    local_path: str
    is_text: bool
    text_content: Optional[str] = None
    content_type: Optional[str] = None


def get_target_upload_dir() -> Path:
    """Résout et crée le répertoire de destination pour les pièces jointes."""
    candidates = [
        os.environ.get("HERMES_UPLOADS_DIR"),
        "/workspace/uploads",
        str(Path.cwd() / "cache" / "uploads"),
        str(Path.cwd() / "uploads"),
    ]
    for c in candidates:
        if not c:
            continue
        try:
            p = Path(c)
            p.mkdir(parents=True, exist_ok=True)
            return p
        except Exception:
            continue
    fallback = Path.cwd() / "uploads"
    fallback.mkdir(parents=True, exist_ok=True)
    return fallback


async def download_and_cache_attachment(
    url: str,
    filename: str,
    size: int,
    content_type: Optional[str] = None,
    session: Optional[aiohttp.ClientSession] = None,
    uploads_dir: Optional[str] = None,
) -> CachedAttachment:
    """Télécharge une pièce jointe Discord, l'enregistre localement et extrait son texte si applicable."""
    upload_dir = Path(uploads_dir) if uploads_dir else get_target_upload_dir()
    upload_dir.mkdir(parents=True, exist_ok=True)
    safe_name = re.sub(r"[^a-zA-Z0-9._-]", "_", filename)
    target_path = upload_dir / f"{int(time.time())}_{safe_name}"

    close_session = False
    if session is None:
        session = aiohttp.ClientSession()
        close_session = True

    try:
        async with session.get(url) as resp:
            if resp.status != 200:
                raise RuntimeError(f"Échec téléchargement ({resp.status}) pour {url}")
            data = await resp.read()
            target_path.write_bytes(data)
    finally:
        if close_session:
            await session.close()

    ext = target_path.suffix.lower()
    is_text = ext in TEXT_EXTENSIONS or (bool(content_type and content_type.startswith("text/")))

    text_content: Optional[str] = None
    if is_text and target_path.stat().st_size <= MAX_INJECT_SIZE_BYTES:
        try:
            text_content = target_path.read_text(encoding="utf-8", errors="replace")
        except Exception as e:
            logger.warning("Impossible d'extraire le texte de %s : %s", filename, e)

    return CachedAttachment(
        name=filename,
        url=url,
        size=target_path.stat().st_size if target_path.exists() else size,
        local_path=str(target_path),
        is_text=is_text,
        text_content=text_content,
        content_type=content_type,
    )
